README.md
# Управление яркостью дисплея с помощью ладони

## Результат:

- распознается жест "кулак", и яркость экрана снижается до значения близкого к 0%
- распознается жест "ладонь", и яркость экрана увеличивается до значения близкого к 100%
- распознается движения ладони, когда она разогнута, вокруг своей оси, и яркость экрана масштабируется

> Распознавание должно поисходить во время видео

## Установка

#### Через `pipenv`

```shell
$ python -m pip install pipenv
$ pipenv shell
$ pipenv install
```

#### Через `pip`

```shell
$ python -m pip install -r requirements.txt
```

### Запуск

```shell
# без sudo яркость не может быть изменена
$ sudo python main.py
```

---
**© Made with ❤️ by arsikurin**


CHANGELOG.md
# Changelog

- 04.12.2021 - анонсированы проектная деятельность и черновой вариант темы будущего проекта
- 05.12.2021 - тема продумана более детально, есть наброски кода
- 06.12.2021 - изучен способ управления якростью дисплея из python кода
- 07.12.2021 - сделан работающий вариант распознавания сжатой ладони, расжатой ладони, движения ладони вокруг своей оси.
  При распознавании сжатой ладони, яркость экрана уменьшается до значения близкого к 0%. При распознавании разжатой
  ладони, яркость экрана увеличивается до значения близкого к 100%. При распознавании движения ладони вокруг своей оси,
  яркость экрана масштабируется.


![wdwdwede](https://media.giphy.com/media/KEHV315CwJfYVtM6as/giphy-downsized.gif)
_Вращение ладони_

<p>
<img width="300" alt="Clenched palm" src="https://user-images.githubusercontent.com/58228813/146271670-75f43476-f886-4c0c-8d17-0eb7b72763fa.png">
<em>Ладонь сжата в кулак</em>
  </p>

<p>
<img width="300" alt="Unclenched palm" src="https://user-images.githubusercontent.com/58228813/146271663-c85f3d5c-28c0-4cb8-89b6-72327001da06.png">
<em>Ладонь разжата</em>
  </p>
 
- 08.12.2021 - сделан туториал
- 15.12.2021 - туториал включается при первом запуске





> Распознавание должно поисходить во время видео
---
**© Made with ❤️ by arsikurin**


stop.py
import collections

import cv2
import mediapipe as mp
import numpy as np

from dependencies import Vector

Results = collections.namedtuple("Results", field_names=("multi_hand_landmarks", "multi_handedness"))


def gen_vectors(img, mcp_num: int, tip_num: int, i):
    mcp = results.multi_hand_landmarks[i].landmark[mcp_num]
    tip = results.multi_hand_landmarks[i].landmark[tip_num]
    cv2.line(
        img,
        (round(mcp.x * img.shape[1]), round(mcp.y * img.shape[0])),
        (round(tip.x * img.shape[1]), round(tip.y * img.shape[0])),
        (255, 0, 0), 5
    )
    return Vector(
        _x=mcp.x * img.shape[1],
        _y=mcp.y * img.shape[0],
        _a=tip.x * img.shape[1],
        _b=tip.y * img.shape[0]
    )


with mp.solutions.hands.Hands(
        static_image_mode=True,
        max_num_hands=1,
        min_detection_confidence=0.5
) as handsDetector:
    cap = cv2.VideoCapture(0)
    while cap.isOpened():
        ret, frame = cap.read()
        if cv2.waitKey(1) & 0xFF == ord("q") or not ret:
            break
        flippedRGB = cv2.cvtColor(np.fliplr(frame), cv2.COLOR_BGR2RGB)

        results: Results = handsDetector.process(flippedRGB)
        if results.multi_hand_landmarks is not None:
            for i, hand in enumerate(results.multi_hand_landmarks):
                thumb_vec = gen_vectors(flippedRGB, 2, 4, i)
                index_vec = gen_vectors(flippedRGB, 5, 8, i)
                middle_vec = gen_vectors(flippedRGB, 9, 12, i)
                ring_vec = gen_vectors(flippedRGB, 13, 16, i)
                pinky_vec = gen_vectors(flippedRGB, 17, 20, i)

                flags = (
                    1.745 > thumb_vec.angle > 1.308,
                    1.745 > index_vec.angle > 1.308,
                    1.745 > middle_vec.angle > 1.308,
                    1.745 > ring_vec.angle > 1.308,
                    1.745 > pinky_vec.angle > 1.308
                )
                if all(flags):
                    print("STOP!")

                # for finger in results.multi_hand_landmarks[i].landmark:
                #     x_tip = int(
                #         finger.x * flippedRGB.shape[1]
                #     )
                #     y_tip = int(
                #         finger.y * flippedRGB.shape[0]
                #     )
                #
                #     cv2.circle(flippedRGB, (x_tip, y_tip), 5, (255, 0, 0), -1)
                # print(results.multi_handedness)
                # print(results.multi_hand_landmarks[i])
        res_image = cv2.cvtColor(flippedRGB, cv2.COLOR_RGB2BGR)
        cv2.imshow("Hands", res_image)


dependencies/vector.py
import math

EPS = 2.220446049250313e-16


class Point:
    def __init__(self, _x, _y=None, _polar=False):
        if isinstance(_x, Point):
            self.polar = _x.polar
            self.angle = _x.angle
            self.x = _x.x
            self.y = _x.y
            self.r = _x.r
        else:
            if _polar:
                self.r = _x
                self.angle = _y
                self.y = self.r * math.sin(self.angle)
                self.x = self.r * math.cos(self.angle)
                self.polar = True
            else:
                self.x = _x
                self.y = _y
                self.r = math.hypot(self.x, self.y)
                if abs(self.r) > EPS:
                    self.angle = math.acos(self.x / self.r)
                else:
                    self.angle = 0
                self.polar = False

    def __str__(self):
        return f"{self.x} {self.y}"

    def dist(self, _x=None, _y=None):
        if _x is None:
            return self.r
        elif _y is None:
            return math.hypot(self.x - _x.x, self.y - _x.y)
        return math.hypot(self.x - _x, self.y - _y)

    def __abs__(self):
        return self.r


class Vector(Point):
    def __init__(self, _x, _y=None, _a=None, _b=None):
        if isinstance(_y, Point):
            super().__init__(_y.x - _x.x, _y.y - _x.y)
        elif isinstance(_x, Point):
            super().__init__(_x.x, _x.y)
        elif isinstance(_b, float):
            super().__init__(_a - _x, _b - _y)
        elif isinstance(_y, float):
            super().__init__(_x, _y)
        elif isinstance(_x, Vector):
            super().__init__(_x.x, _y.x)

    def dot_product(self, other):
        return self.x * other.x + self.y * other.y

    def cross_product(self, other):
        return self.x * other.y - self.y * other.x

    def mul(self, other):
        return Vector(self.x * other, self.y * other)


main.py
import collections
import math

import cv2
import mediapipe as mp
import numpy as np
import screen_brightness_control as sbc

Results = collections.namedtuple("Results", field_names=("multi_hand_landmarks", "multi_handedness"))


def get_points(landmark, shape):
    points = []
    for mark in landmark:
        points.append((mark.x * shape[1], mark.y * shape[0]))
    return np.array(points, dtype=np.int32)


def palm_size(landmark, shape):
    """
    :return: palm enclosing circle radius
    """

    x1, y1 = landmark[0].x * shape[1], landmark[0].y * shape[0]
    x2, y2 = landmark[5].x * shape[1], landmark[5].y * shape[0]
    return math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2)


def help_():
    with mp.solutions.hands.Hands(
            static_image_mode=False,
            max_num_hands=1,
            min_detection_confidence=.2
    ) as handsDetector:
        flags = [0, 0, 0]
        cap = cv2.VideoCapture(0)
        prev_fist = False
        while cap.isOpened():
            ret, frame = cap.read()
            if cv2.waitKey(1) & 0xFF == ord("q") or not ret:
                break
            flipped = np.fliplr(frame)
            flippedRGB = cv2.cvtColor(flipped, code=cv2.COLOR_BGR2RGB)
            results: Results = handsDetector.process(flippedRGB)
            if results.multi_hand_landmarks is not None:
                cv2.circle(
                    flippedRGB,
                    center=(int(results.multi_hand_landmarks[0].landmark[4].x * flippedRGB.shape[1]),
                            int(results.multi_hand_landmarks[0].landmark[4].y * flippedRGB.shape[0])),
                    radius=5, color=(255, 0, 0), thickness=-1
                )
                cv2.circle(
                    flippedRGB, center=(int(results.multi_hand_landmarks[0].landmark[20].x * flippedRGB.shape[1]),
                                        int(results.multi_hand_landmarks[0].landmark[20].y * flippedRGB.shape[0])),
                    radius=5, color=(255, 0, 0), thickness=-1
                )
                cv2.line(
                    flippedRGB,
                    pt1=(int(results.multi_hand_landmarks[0].landmark[4].x * flippedRGB.shape[1]),
                         int(results.multi_hand_landmarks[0].landmark[4].y * flippedRGB.shape[0])),
                    pt2=(int(results.multi_hand_landmarks[0].landmark[20].x * flippedRGB.shape[1]),
                         int(results.multi_hand_landmarks[0].landmark[20].y * flippedRGB.shape[0])),
                    color=(255, 0, 0), thickness=2
                )

                (x, y), r = cv2.minEnclosingCircle(
                    get_points(results.multi_hand_landmarks[0].landmark, flippedRGB.shape))
                ws = palm_size(results.multi_hand_landmarks[0].landmark, flippedRGB.shape)

                if flags[0] < 7:
                    cv2.putText(
                        flippedRGB, text=f"Try clenching your fist ({flags[0]} / 6)",
                        org=(10, flippedRGB.shape[0] - 10),
                        fontFace=cv2.FONT_HERSHEY_DUPLEX, fontScale=1, color=(255, 255, 255), thickness=1
                    )
                    if 2 * r / ws > 1.6:
                        cv2.circle(flippedRGB, center=(int(x), int(y)), radius=int(r), color=(0, 0, 255), thickness=2)
                        if prev_fist:
                            prev_fist = False
                            flags[0] += 1
                        cv2.putText(
                            flippedRGB, text="status: [PALM]", org=(10, 50), fontFace=cv2.FONT_HERSHEY_PLAIN,
                            fontScale=3, color=(255, 255, 255), thickness=2
                        )
                    else:
                        cv2.circle(flippedRGB, (int(x), int(y)), int(r), (0, 255, 0), 2)
                        if not prev_fist:
                            prev_fist = True
                            flags[0] += 1
                        cv2.putText(
                            flippedRGB, text="status: [FIST]", org=(10, 50), fontFace=cv2.FONT_HERSHEY_PLAIN,
                            fontScale=3, color=(255, 255, 255), thickness=2
                        )

                elif flags[1] < 7:
                    cv2.putText(
                        flippedRGB, text="Try rotating your unclenched fist", org=(10, flippedRGB.shape[0] - 40),
                        fontFace=cv2.FONT_HERSHEY_DUPLEX, fontScale=1, color=(255, 255, 255), thickness=1
                    )
                    cv2.putText(
                        flippedRGB, text=f"around axis of the ordinate ({flags[1]} / 6)",
                        org=(10, flippedRGB.shape[0] - 10),
                        fontFace=cv2.FONT_HERSHEY_DUPLEX, fontScale=1, color=(255, 255, 255), thickness=1
                    )
                    if 2 * r / ws > 1.6:
                        cv2.circle(flippedRGB, center=(int(x), int(y)), radius=int(r), color=(0, 0, 255), thickness=2)
                        pinky_tip = results.multi_hand_landmarks[0].landmark[20]
                        thumb_tip = results.multi_hand_landmarks[0].landmark[4]
                        if (pinky_tip.x - thumb_tip.x + .2) * 200 > 60 and flags[2] == 0:
                            flags[1] += 1
                            flags[2] = 1
                        elif (pinky_tip.x - thumb_tip.x + .2) * 200 < 20 and flags[2] == 1:
                            flags[1] += 1
                            flags[2] = 0

                        cv2.putText(
                            flippedRGB, text="status: [ROTATING]", org=(10, 50), fontFace=cv2.FONT_HERSHEY_PLAIN,
                            fontScale=3, color=(255, 255, 255), thickness=2
                        )
                else:
                    with open("service.arsikurin", "w") as f:
                        f.write("1")
                    cv2.putText(
                        flippedRGB, text=f"Congrats! You've completed tutorial.", org=(10, flippedRGB.shape[0] - 40),
                        fontFace=cv2.FONT_HERSHEY_DUPLEX, fontScale=1, color=(255, 255, 255), thickness=1
                    )
                    cv2.putText(
                        flippedRGB, text=f"You should rerun the script", org=(10, flippedRGB.shape[0] - 10),
                        fontFace=cv2.FONT_HERSHEY_DUPLEX, fontScale=1, color=(255, 255, 255), thickness=1
                    )

            res_image = cv2.cvtColor(flippedRGB, code=cv2.COLOR_RGB2BGR)
            cv2.imshow("Result", res_image)


def main():
    with mp.solutions.hands.Hands(
            static_image_mode=True,
            max_num_hands=1,
            min_detection_confidence=.2
    ) as handsDetector:
        cap = cv2.VideoCapture(0)
        prev_fist = False
        while cap.isOpened():
            ret, frame = cap.read()
            if cv2.waitKey(1) & 0xFF == ord("q") or not ret:
                break
            flipped = np.fliplr(frame)
            flippedRGB = cv2.cvtColor(flipped, code=cv2.COLOR_BGR2RGB)
            results: Results = handsDetector.process(flippedRGB)

            if results.multi_hand_landmarks is not None:
                cv2.circle(
                    flippedRGB,
                    center=(int(results.multi_hand_landmarks[0].landmark[4].x * flippedRGB.shape[1]),
                            int(results.multi_hand_landmarks[0].landmark[4].y * flippedRGB.shape[0])),
                    radius=5, color=(255, 0, 0), thickness=-1
                )
                cv2.circle(
                    flippedRGB, center=(int(results.multi_hand_landmarks[0].landmark[20].x * flippedRGB.shape[1]),
                                        int(results.multi_hand_landmarks[0].landmark[20].y * flippedRGB.shape[0])),
                    radius=5, color=(255, 0, 0), thickness=-1
                )
                cv2.line(
                    flippedRGB,
                    pt1=(int(results.multi_hand_landmarks[0].landmark[4].x * flippedRGB.shape[1]),
                         int(results.multi_hand_landmarks[0].landmark[4].y * flippedRGB.shape[0])),
                    pt2=(int(results.multi_hand_landmarks[0].landmark[20].x * flippedRGB.shape[1]),
                         int(results.multi_hand_landmarks[0].landmark[20].y * flippedRGB.shape[0])),
                    color=(255, 0, 0), thickness=2
                )

                (x, y), r = cv2.minEnclosingCircle(
                    get_points(results.multi_hand_landmarks[0].landmark, flippedRGB.shape))
                ws = palm_size(results.multi_hand_landmarks[0].landmark, flippedRGB.shape)
                if 2 * r / ws > 1.6:
                    cv2.circle(flippedRGB, center=(int(x), int(y)), radius=int(r), color=(0, 0, 255), thickness=2)
                    if prev_fist:
                        sbc.fade_brightness(finish=100, start=1, blocking=False, increment=10)
                        prev_fist = False
                        cv2.putText(
                            flippedRGB, text="status: [PALM]", org=(10, 50), fontFace=cv2.FONT_HERSHEY_PLAIN,
                            fontScale=3, color=(255, 255, 255), thickness=2
                        )
                    else:
                        pinky_tip = results.multi_hand_landmarks[0].landmark[20]
                        thumb_tip = results.multi_hand_landmarks[0].landmark[4]
                        sbc.set_brightness((pinky_tip.x - thumb_tip.x + .2) * 200)
                        cv2.putText(
                            flippedRGB, text="status: [ROTATING]", org=(10, 50), fontFace=cv2.FONT_HERSHEY_PLAIN,
                            fontScale=3, color=(255, 255, 255), thickness=2
                        )
                else:
                    cv2.circle(flippedRGB, (int(x), int(y)), int(r), (0, 255, 0), 2)
                    if not prev_fist:
                        sbc.fade_brightness(finish=1, start=100, blocking=False, increment=10)
                        prev_fist = True
                    cv2.putText(
                        flippedRGB, text="status: [FIST]", org=(10, 50), fontFace=cv2.FONT_HERSHEY_PLAIN,
                        fontScale=3, color=(255, 255, 255), thickness=2
                    )

            res_image = cv2.cvtColor(flippedRGB, code=cv2.COLOR_RGB2BGR)
            cv2.imshow("Result", res_image)


if __name__ == "__main__":
    try:
        with open("service.arsikurin", "r") as f:
            if f.read() == "1":
                main()
            else:
                help_()
    except FileNotFoundError:
        help_()
